### PR TITLE
CORE-3856 - kafka max allowed msg size config

### PR DESCRIPTION
**Pull request by driessamyn**
Take Max Allowed Kafka Message size from config.

Related API PR: https://github.com/corda/corda-api/pull/412

### CODE REVIEW

File: applications/tools/flow-worker-setup/src/main/kotlin/net/corda/applications/flowworker/setup/tasks/SetupVirtualNode.kt [hunk 1]
```diff
@@ -60,7 +59,7 @@ class SetupVirtualNode(private val context: TaskContext) : Task {
 
         cpiList.flatMap { it.cpks }.map { cpk ->
             val cpkChecksum = cpk.metadata.fileChecksum
-            val chunkWriter = ChunkWriterFactory.create(SUGGESTED_CHUNK_SIZE)
+            val chunkWriter = ChunkWriterFactory.create(972800)
```
Comment (barrylapthorn) [inline applications/tools/flow-worker-setup/src/main/kotlin/net/corda/applications/flowworker/setup/tasks/SetupVirtualNode.kt:62]: What's the magic number?  Could it be a static?

File: components/virtual-node/cpk-write-service-impl/src/main/kotlin/net/corda/cpk/write/impl/CpkWriteServiceImpl.kt [hunk 2]
```diff
@@ -150,22 +153,23 @@ class CpkWriteServiceImpl @Activate constructor(
      * We've received a config event that we care about, we can now write cpks
      */
     private fun onConfigChangedEvent(event: ConfigChangedEvent, coordinator: LifecycleCoordinator) {
-        val config = event.config.getConfig(MESSAGING_CONFIG)
+        val messagingConfig = event.config.getConfig(MESSAGING_CONFIG)
```
Comment (barrylapthorn) [inline components/virtual-node/cpk-write-service-impl/src/main/kotlin/net/corda/cpk/write/impl/CpkWriteServiceImpl.kt:156]: I tend to leave renames like this alone as it results in a smaller diff that's less "noisy".

File: components/virtual-node/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/EntityProcessorFactoryImpl.kt [hunk 3]
```diff
@@ -29,27 +30,31 @@ class EntityProcessorFactoryImpl @Activate constructor(
...
+                log.debug("Payload $kb kb < max Kafka payload size ${maxPayloadSize / (1024)} kb")
             return bytes
         }
     }
 
     override fun create(config: SmartConfig): FlowPersistenceProcessor {
         val subscriptionConfig = SubscriptionConfig(GROUP_NAME, Schemas.VirtualNode.ENTITY_PROCESSOR)
-        val processor = EntityMessageProcessor(entitySandboxService, UTCClock(), EntityProcessorFactoryImpl::payloadSizeCheck)
+        // max allowed msg size minus headroom for wrapper message
+        val maxPayLoadSize = config.getInt(MessagingConfig.MAX_ALLOWED_MSG_SIZE) - 1024
```
Comment (barrylapthorn) [inline components/virtual-node/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/EntityProcessorFactoryImpl.kt]: What's the `1024`?  Can we name that as `CORDA_MESSAGE_OVERHEAD` or something?

File: libs/chunking/chunking-core/src/main/kotlin/net/corda/chunking/impl/ChunkWriterImpl.kt [hunk 4]
```diff
@@ -17,11 +17,13 @@ import java.util.UUID
 /**
  * Chunks up a binary into smaller parts and passes them to the supplied callback.
  */
-internal class ChunkWriterImpl(val chunkSize: Int, private val properties: Map<String, String?>? = null) : ChunkWriter {
+internal class ChunkWriterImpl(val maxAllowedMessageSize: Int, private val properties: Map<String, String?>? = null) : ChunkWriter {
     companion object {
         const val KB = 1024
         const val MB = 1024 * KB
 
+        const val BUFFER_SIZE = 1024 * 10
```
Comment (barrylapthorn) [inline libs/chunking/chunking-core/src/main/kotlin/net/corda/chunking/impl/ChunkWriterImpl.kt]: I think I'd prefer `CORDA_MESSAGE_OVERHEAD` (_bytes?) or something.

File: libs/virtual-node/cpi-upload-manager-impl/src/main/kotlin/net/corda/libs/cpiupload/impl/CpiUploadManagerFactoryImpl.kt [hunk 5]
```diff
@@ -47,18 +48,23 @@ class CpiUploadManagerFactoryImpl : CpiUploadManagerFactory {
     }
 
     override fun create(
-        config: SmartConfig,
+        messagingConfig: SmartConfig,
```
Comment (barrylapthorn) [inline libs/virtual-node/cpi-upload-manager-impl/src/main/kotlin/net/corda/libs/cpiupload/impl/CpiUploadManagerFactoryImpl.kt:51]: It doesn't feel like renaming adds much here other than a larger diff.

**Review by barrylapthorn — DISMISSED** [review]
Seems reasonable - we expect the max (kafka) message size in the config.

I'd just convert some magic numbers to named constants for readability / self-documenting.

File: applications/tools/flow-worker-setup/src/main/kotlin/net/corda/applications/flowworker/setup/tasks/SetupVirtualNode.kt (hunk 1 above)
Comment (driessamyn) [reply applications/tools/flow-worker-setup/src/main/kotlin/net/corda/applications/flowworker/setup/tasks/SetupVirtualNode.kt:62]: It's a test tool. Not sure where it should come from.

File: components/virtual-node/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/EntityProcessorFactoryImpl.kt (hunk 3 above)
Comment (driessamyn) [reply components/virtual-node/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/EntityProcessorFactoryImpl.kt]: Y. It's fairly arbitrary. But good suggestion, I'll put it in a constant.

**Comment by driessamyn** [comment]
> Seems reasonable - we expect the max (kafka) message size in the config.
> 
> I'd just convert some magic numbers to named constants for readability / self-documenting.

Pushed a change.
I did not change back messageConfig -> config as it would probably be more work compared to the review overhead.
I've been trying to change instance of this as it's confusing, especially when there is other (non-messaging) config.

**Review by barrylapthorn — DISMISSED** [review]
LGTM

File: libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/config/KafkaConfigMergerImpl.kt [hunk 6]
```diff
@@ -24,6 +24,7 @@ class KafkaConfigMergerImpl : BusConfigMerger {
         var updatedMessagingConfig = (messagingConfig?: getBaseKafkaMessagingConfig())
             .withValue(BootConfig.INSTANCE_ID, ConfigValueFactory.fromAnyRef(bootConfig.getString(BootConfig.INSTANCE_ID)))
             .withValue(BootConfig.TOPIC_PREFIX, ConfigValueFactory.fromAnyRef(bootConfig.getString(BootConfig.TOPIC_PREFIX)))
+            .withValue(BUS_TYPE, ConfigValueFactory.fromAnyRef("KAFKA"))
```
Comment (driessamyn) [inline libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/config/KafkaConfigMergerImpl.kt:27]: @LWogan , @JamesHR3 

This change was suggested by Lorcan and looks like it fixes the issue with the BUS section not being merged. 
This associated corda-api change is required: https://github.com/corda/corda-api/pull/417

Comment (LWogan) [reply libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/config/KafkaConfigMergerImpl.kt:27]: I think you can just get rid of getBaseKafkaMessagingConfig and just always populate the busType as you have done. 

Then this line can just be the below or something equivalent 
`var updatedMessagingConfig = (messagingConfig?: SmartConfigImpl.empty()) `

Comment (JamesHR3) [reply libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/config/KafkaConfigMergerImpl.kt:27]: Seems reasonable to me.

Comment (driessamyn) [reply libs/messaging/kafka-message-bus-impl/src/main/kotlin/net/corda/messagebus/kafka/config/KafkaConfigMergerImpl.kt:27]: @LWogan , I tried this change, but it fails as `InstanceId` doesn't get defaulted and hence is missing from the published config:

```
[pod/corda-crypto-worker-8df85cd87-p9cpq/corda-crypto-worker] 12:59:03.896 [lifecycle-coordinator-0] ERROR net.corda.messaging.config.MessagingConfigResolver - Failed to resolve publisher config PublisherConfig(clientId=crypto.processor, transactional=true): merge of empty config,bundle://5d2f1b72-b98b-459e-a110-534893d15346_15.0:1/messaging-defaults.conf: 1: No configuration setting found for key 'instanceId'
```

Maybe this is a problem specific to the Crypto worker, but I'd prefer to not include a fix for it in this PR.

**Review by barrylapthorn — APPROVED** [review]

**Pull request merged**